moveit_ros/hybrid_planning/hybrid_planning_manager/hybrid_planning_manager_component/include/moveit/hybrid_planning_manager/hybrid_planning_manager.h

```diff
@@ -113,6 +120,9 @@ class HybridPlanningManager : public rclcpp::Node
   // Flag that indicates whether the manager is initialized
   bool initialized_;
 
+  // Flag that indicates hybrid planning has been canceled
+  bool stop_hybrid_planning_;
```

Comment:
Is this being used to communicate between concurrent sections of code?  If so, please use a lock-free queue or other message-passing technique instead of shared memory.

Here is a lock-free queue: https://www.boost.org/doc/libs/1_76_0/doc/html/boost/lockfree/spsc_queue.html

For reasoning read this blog post: http://henrikeichenhardt.blogspot.com/2013/06/why-shared-mutable-state-is-root-of-all.html

For examples from other languages see this section from the Rust Book: https://doc.rust-lang.org/book/ch16-02-message-passing.html and this section from the Golang documentation: https://go.dev/doc/effective_go#concurrency

For an alternate solution, you could use strong typing and have the code divided between behavior for executing and not executing and change a type that determines the action being taken.  Then you could use RAII for cleaning up and transitioning between your two states.

Reply:
Don't you need to know how many threads are talking to use spsc queue? I don't know that here. We could probably assume 2 but...

Reply:
yes, you do... you need one producer and one consumer thread or it is not thread safe.  if you have multiple consumers of this you should use a multi-producer multi-consumer queue: https://www.boost.org/doc/libs/1_59_0/doc/html/boost/lockfree/queue.html

Reply:
If you do have multiple consumers you'll need to know how many to make sure you notify each.  It might be better to have a stack of spsc queues that consumers can register themselves into.

Reply:
So how many producers and consumers do you think there are? I'm not trying to be difficult here, but that number needs to be known. I think it depends on whether a SingleThreadedExecutor or MultiThreadedExecutor is being used. MultiThreadedExecutor [defaults](https://github.com/ros2/rclcpp/issues/441) to match (num. hardware threads), it looks like. None of that stuff is determined in this header file.

Reply:
I don't know what the right answer is in this case.  Another question I have is are these callbacks re-entrant and should we guard against that too.  Another way to ask that question is what would happen if on two separate threads ROS calls one of these callbacks twice at the same time?

For now, go ahead and use some sort of synchronization.  Long term we need a better model for how to approach this than shared memory if we want MoveIt to scale onto many threads well.